asynch/pool.py

```diff
@@ -140,26 +140,46 @@ async def _create_connection(self) -> None:
         if pool_size > maxsize:
             raise AsynchPoolError(f"{self} is overburden")
 
-        conn = await connect(**self._connection_kwargs)
+        conn = Connection(**self._connection_kwargs)
+        await conn.connect()
         self._free_connections.append(conn)
 
-    async def _acquire_connection(self) -> Connection:
+    def _pop_connection(self) -> Connection:
         if not self._free_connections:
             raise AsynchPoolError(f"no free connection in {self}")
 
-        conn = self._free_connections.popleft()
+        return self._free_connections.popleft()
+
+    async def _acquire_connection(self) -> Connection:
+        conn = self._pop_connection()
+
+        try:
+            await conn._refresh()
+        except ConnectionError:
+            await self._create_connection()
+            conn = self._pop_connection()
+
         self._acquired_connections.append(conn)
         return conn
 
     async def _release_connection(self, conn: Connection) -> None:
         if conn not in self._acquired_connections:
             raise AsynchPoolError(f"the connection {conn} does not belong to {self}")
 
+        to_recreate: bool = False
+        try:
+            await conn._refresh()
+            self._free_connections.append(conn)
+        except ConnectionError:
+            to_recreate = True
+
         self._acquired_connections.remove(conn)
-        self._free_connections.append(conn)
+        if to_recreate:
+            await self._create_connection()
```

Comment:
Checking if the given connection is fresh (alive). If not, the invalid connection will not fill the ranks of the available  `free_connections` deque because of the raised `ConnectionError`. The exception is intercepted and the flag `to_recreate` (a connection) is set to True and this procedure will be done only after removing the previously acquired connection from the registry of `acquired_connections` for the consistency sake, otherwise `self.connections` in the `_init_connections` method will result in miscounted number.

Reply:
I'm not sure if we're not overcomplicating this. My approach would be:

1. Get connections from the pool.
2. Check if it's still open. If not goto1.
3. If the pool is empty, create a new connection.

Reply:
Redone.